bot/utils/redis_cache.py

```diff
@@ -0,0 +1,360 @@
+        if self._namespace is None:
+            log.exception("Attempt to use RedisCache with no namespace.")
+            raise RuntimeError(
+                "Critical error: RedisCache has no namespace. "
+                "Did you initialize this object as a class attribute?"
+            )
+        await self.bot.redis_ready.wait()
+
+    def __set_name__(self, owner: Any, attribute_name: str) -> None:
+        """
+        Set the namespace to Class.attribute_name.
+
+        Called automatically when this class is constructed inside a class as an attribute.
+
+        This class MUST be created as a class attribute in a class, otherwise it will raise
+        exceptions whenever a method is used. This is because it uses this method to create
+        a namespace like `MyCog.my_class_attribute` which is used as a hash name when we store
+        stuff in Redis, to prevent collisions.
+        """
+        self._set_namespace(f"{owner.__name__}.{attribute_name}")
+
+    def __get__(self, instance: RedisCache, owner: Any) -> RedisCache:
+        """
+        This is called if the RedisCache is a class attribute, and is accessed.
+
+        The class this object is instantiated in must contain an attribute with an
+        instance of Bot. This is because Bot contains our redis_session, which is
+        the mechanism by which we will communicate with the Redis server.
+
+        Any attempt to use RedisCache in a class that does not have a Bot instance
+        will fail. It is mostly intended to be used inside of a Cog, although theoretically
+        it should work in any class that has a Bot instance.
+        """
+        if self.bot:
+            return self
+
+        if self._namespace is None:
+            log.exception("RedisCache must be a class attribute.")
+            raise RuntimeError("RedisCache must be a class attribute.")
+
+        if instance is None:
+            log.exception(
+                "Attempt to access RedisCache instance through the cog's class object "
+                "before accessing it through the cog instance."
+            )
+            raise RuntimeError(
+                "You must access the RedisCache instance through the cog instance "
+                "before accessing it using the cog's class object."
+            )
+
+        for attribute in vars(instance).values():
+            if isinstance(attribute, Bot):
+                self.bot = attribute
+                self._redis = self.bot.redis_session
+                return self
+        else:
+            log.exception("Attempt to use RedisCache with no `Bot` instance.")
+            raise RuntimeError(
+                "Critical error: RedisCache has no `Bot` instance. "
+                "This happens when the class RedisCache was created in doesn't "
+                "have a Bot instance. Please make sure that you're instantiating "
+                "the RedisCache inside a class that has a Bot instance "
+                "class attribute."
```

Comment:
Let's drop "class" cause it may imply to some that instance attributes aren't acceptable.
```suggestion
                "attribute."
```